exporters/etw/include/opentelemetry/exporters/etw/etw_provider_exporter.h

```diff
@@ -0,0 +1,664 @@
+      case EventFormat::ETW_MSGPACK:
+      {
+        REGHANDLE hProvider = 0;
+        if (EventRegister(&data.providerGuid, NULL, NULL, &hProvider) != ERROR_SUCCESS)
+        {
+          // There was an error registering the ETW provider
+          data.providerHandle = INVALID_HANDLE;
+        }
+        else
+        {
+          data.providerHandle = hProvider;
+        }
+      };
+      break;
+#endif
+
+      default:
+        // TODO: other protocols, e.g. XML events - not supported yet
+        break;
+    }
+
+    // We always return an entry even if we failed to register.
+    // Caller should check whether the hProvider handle is valid.
+    return data;
+  }
+
+  /// <summary>
+  /// Unregister Provider
+  /// </summary>
+  /// <param name="providerId"></param>
+  /// <returns></returns>
+  unsigned long close(Handle data)
+  {
+    std::lock_guard<std::mutex> lock(m_providerMapLock);
+
+    auto m  = providers();
+    auto it = m.begin();
+    while (it != m.end())
+    {
+      if (it->second.providerHandle == data.providerHandle)
+      {
+        auto result = EventUnregister(data.providerHandle);
+        m.erase(it);
+        return result;
+      }
+    };
+    return STATUS_ERROR;
+  }
+
+#ifdef HAVE_MSGPACK
+  template <class T>
+  unsigned long writeMsgPack(Handle &providerData, T eventData)
+  {
+
+    // Make sure you stop sending event before register unregistering providerData
+    if (providerData.providerHandle == INVALID_HANDLE)
+    {
+      // Provider not registered!
+      return STATUS_ERROR;
+    };
+
+    const std::string EVENT_NAME = "name";
+    std::string eventName        = "NoName";
+    auto nameField               = eventData[EVENT_NAME];
+    switch (nameField.index())
+    {
+      case common::AttributeType::TYPE_STRING:
+        eventName =
+            (char *)(nostd::get<nostd::string_view>(nameField).data());  // must be 0-terminated!
+        break;
+      case common::AttributeType::TYPE_CSTRING:
+        eventName = (char *)(nostd::get<const char *>(nameField));
+        break;
+      default:
+        // This is user error. Invalid event name!
+        // We supply default 'NoName' event name in this case.
+        break;
+    }
+
+    /* clang-format off */
+    nlohmann::json jObj =
+    {
+      { "env_name", "Span" },
+      { "env_ver", "4.0" },
+
+      // TODO: remove these fields. Provided for illustrative purposes only.
+      { "env_cloud_role", "BusyWorker" },
+      { "env_cloud_roleInstance", "CY1SCH030021417" },
+      { "env_cloud_roleVer", "9.0.15289.2" },
+      //
+
+      // TODO: compute time in MessagePack-friendly format
+      // TODO: should we consider uint64_t format with Unix timestamps for ELK stack?
+      { "env_time",
+        {
+          { "TypeCode", 255 },
+          { "Body","0xFFFFFC60000000005F752C2C" }
+        }
+      },
+      //
+
+      // TODO: follow JSON implementation of OTLP or place protobuf for non-Microsoft flows
```

Comment:
Here we need to understand what are the attribute names for `common` fields:
- name
- time (as in when `event` got logged)
- traceId
- spanId
- (span?) start time
- (span?) end time

Introduce a set of macros that would define the attribute names. Use these macros, then we can rename these. The other alternative is to use `const char *` - set of pooled constants in some namespace, e.g.
```cpp
    nlohmann::json jObj =
    {
      { OPENTELEMETRY_SPAN_NAME , "Span" },
...
```

Reply:
https://github.com/open-telemetry/opentelemetry-cpp/issues/466

Reply:
Let's solve this in https://github.com/open-telemetry/opentelemetry-cpp/issues/466

Reply:
I think independent from the possibility for vendor specific solutions, one should also implement and provide a standard encoding. This should be documented, ideally in a README provided with the exporter.